### PR TITLE
turtlesim for ROS2

### PR DESCRIPTION
**Pull request by routiful**
I have been ported turtlesim node(including tutorials) to ROS2.
It works same as ROS but has some differences in tutorials.

Add some information when mimic node is started
(https://github.com/ROBOTIS-Platform/ros_tutorials/blob/f8b600d331060f39afc44436b3016d99a9fead97/turtlesim/tutorials/mimic.cpp#L15)

Add argument to add namespace in mimic node
(https://github.com/ROBOTIS-Platform/ros_tutorials/blob/f8b600d331060f39afc44436b3016d99a9fead97/turtlesim/tutorials/mimic.cpp#L64)

Add stop signal in turtle_teleop node
(https://github.com/ROBOTIS-Platform/ros_tutorials/blob/f8b600d331060f39afc44436b3016d99a9fead97/turtlesim/tutorials/teleop_turtle_key.cpp#L117)

Please review this and feel free to comment for anything.

### CODE REVIEW

File: turtlesim/src/turtle.cpp [hunk 1]
```diff
@@ -143,12 +161,13 @@ bool Turtle::update(double dt, QPainter& path_painter, const QImage& path_image,
...
   teleport_requests_.clear();
 
-  if (ros::WallTime::now() - last_command_time_ > ros::WallDuration(1.0))
-  {
-    lin_vel_ = 0.0;
-    ang_vel_ = 0.0;
-  }
-
+  // TODO: It makes error
+  // if ((nh_->now() - last_command_time_) > rclcpp::Duration(1.0))
```
Comment (mlanting) [inline turtlesim/src/turtle.cpp]: I ran into this issue - the constructor for rclcpp::Duration() with one argument expects the time in nanoseconds. It has a second constructor that takes two arguments, the first being seconds, and the second being nanoseconds. I made the following change in my code to get it to work:

`if (nh_->now() - last_command_time_ > rclcpp::Duration(1.0, 0))`

Comment (routiful) [reply turtlesim/src/turtle.cpp]: Thank you for your comments! I modified that.

**Comment by routiful** [comment]
Carefully Ping
I assured that turtlesim is the best simulation for ROS2 users.

File: turtlesim/CMakeLists.txt [hunk 2]
```diff
@@ -1,55 +1,110 @@
...
-find_package(catkin REQUIRED COMPONENTS geometry_msgs message_generation rosconsole roscpp roscpp_serialization roslib rostime std_msgs std_srvs)
+if(NOT CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 14)
+endif()
 
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic -Wvla)
+endif()
+
+find_package(ament_cmake REQUIRED)
```
Comment (dirk-thomas) [inline turtlesim/CMakeLists.txt]: Please remove trailing spaces from all modified lines.

File: turtlesim/CMakeLists.txt [hunk 3]
```diff
@@ -1,55 +1,110 @@
...
+  include 
+  ${QT_INCLUDES}
+)
 
 set(turtlesim_node_SRCS
   src/turtlesim.cpp
   src/turtle.cpp
   src/turtle_frame.cpp
 )
+
```
Comment (dirk-thomas) [inline turtlesim/CMakeLists.txt]: Please avoid any unrelated (white space) changes since they are an unnecessary diff between the default branch and the target branch.

File: turtlesim/CMakeLists.txt [hunk 4]
```diff
@@ -1,55 +1,110 @@
...
+rosidl_target_interfaces(mimic
+  ${PROJECT_NAME} "rosidl_typesupport_cpp"
+)
+
+install(TARGETS 
+  turtlesim_node 
+  turtle_teleop_key 
+  draw_square 
+  mimic
```
Comment (dirk-thomas) [inline turtlesim/CMakeLists.txt]: Please keep the original format of this line to avoid introducing unnecessary diffs:

> `install(TARGETS turtlesim_node turtle_teleop_key draw_square mimic`

File: turtlesim/CMakeLists.txt [hunk 5]
```diff
@@ -1,55 +1,110 @@
...
+  DESTINATION lib/${PROJECT_NAME}
+) 
+
+install(DIRECTORY 
+  images
+  launch
```
Comment (dirk-thomas) [inline turtlesim/CMakeLists.txt]: Please only install the `images` (using the previous lines of code). The launch file wasn't installed before.

If you would like the launch file to be installed please contribute that change to the default branch first.

File: turtlesim/include/turtlesim/turtle.h [hunk 6]
```diff
@@ -32,15 +32,18 @@
...
-# include <boost/shared_ptr.hpp>
-
-# include <turtlesim/Pose.h>
-# include <geometry_msgs/Twist.h>
-# include <turtlesim/SetPen.h>
-# include <turtlesim/TeleportRelative.h>
-# include <turtlesim/TeleportAbsolute.h>
-# include <turtlesim/Color.h>
+# include <rclcpp/rclcpp.hpp>
+// #include <boost/shared_ptr.hpp>
```
Comment (dirk-thomas) [inline turtlesim/include/turtlesim/turtle.h]: Please remove such lines instead of commenting them out.

File: turtlesim/include/turtlesim/turtle.h [hunk 7]
```diff
@@ -55,61 +58,59 @@ namespace turtlesim
 
 class Turtle
 {
-public:
-  Turtle(const ros::NodeHandle& nh, const QImage& turtle_image, const QPointF& pos, float orient);
+ public:
+  Turtle(rclcpp::Node::SharedPtr &node_handle, std::string &real_name, const QImage& turtle_image, const QPointF& pos, float orient);
 
-  bool update(double dt, QPainter& path_painter, const QImage& path_image, qreal canvas_width, qreal canvas_height);
+  bool update(double dt, QPainter& path_painter, const QImage& path_image, float canvas_width, float canvas_height);
```
Comment (dirk-thomas) [inline turtlesim/include/turtlesim/turtle.h]: What is the reason for changing these types? It should be possible to leave them as is.

File: turtlesim/include/turtlesim/turtle.h [hunk 8]
```diff
@@ -55,61 +58,59 @@ namespace turtlesim
 
 class Turtle
 {
-public:
-  Turtle(const ros::NodeHandle& nh, const QImage& turtle_image, const QPointF& pos, float orient);
+ public:
```
Comment (dirk-thomas) [inline turtlesim/include/turtlesim/turtle.h]: Please don't change the indentation but keep the style of the code as is.

File: turtlesim/include/turtlesim/turtle.h [hunk 9]
```diff
@@ -55,61 +58,59 @@ namespace turtlesim
...
+  Turtle(rclcpp::Node::SharedPtr &node_handle, std::string &real_name, const QImage& turtle_image, const QPointF& pos, float orient);
```
Comment (dirk-thomas) [inline turtlesim/include/turtlesim/turtle.h]: Please follow the style of the existing code. The ampersand should be on the type not the variable name.

File: turtlesim/include/turtlesim/turtle_frame.h [hunk 10]
```diff
@@ -70,24 +71,20 @@ private slots:
   void clear();
   bool hasTurtle(const std::string& name);
 
-  bool clearCallback(std_srvs::Empty::Request&, std_srvs::Empty::Response&);
-  bool resetCallback(std_srvs::Empty::Request&, std_srvs::Empty::Response&);
-  bool spawnCallback(turtlesim::Spawn::Request&, turtlesim::Spawn::Response&);
-  bool killCallback(turtlesim::Kill::Request&, turtlesim::Kill::Response&);
```
Comment (dirk-thomas) [inline turtlesim/include/turtlesim/turtle_frame.h:76]: Is there a reason to replace these callback methods with lambdas? If not, please keep them as they were before tin order to not introduce unnecessary diffs.

File: turtlesim/src/turtle_frame.cpp [hunk 11]
```diff
@@ -31,38 +31,41 @@
...
 
-TurtleFrame::TurtleFrame(QWidget* parent, Qt::WindowFlags f)
-: QFrame(parent, f)
-, path_image_(500, 500, QImage::Format_ARGB32)
-, path_painter_(&path_image_)
-, frame_count_(0)
-, id_counter_(0)
+TurtleFrame::TurtleFrame(rclcpp::Node::SharedPtr &node_handle, QWidget* parent, Qt::WindowFlags f)
+: QFrame(parent, f),
+  path_image_(500, 500, QImage::Format_ARGB32),
```
Comment (dirk-thomas) [inline turtlesim/src/turtle_frame.cpp]: Please keep the style of the commas as is. There is no need to change them.

File: turtlesim/src/turtle_frame.cpp [hunk 12]
```diff
@@ -31,38 +31,41 @@
...
+  id_counter_(0)
 {
   setFixedSize(500, 500);
   setWindowTitle("TurtleSim");
 
   srand(time(NULL));
 
   update_timer_ = new QTimer(this);
-  update_timer_->setInterval(16);
+  update_timer_->setInterval(16); // msec
```
Comment (dirk-thomas) [inline turtlesim/src/turtle_frame.cpp]: Since this is unrelated to ROS 2 please consider contributing this against the default branch instead.

File: turtlesim/tutorials/teleop_turtle_key.cpp [hunk 13]
```diff
@@ -81,55 +80,58 @@ void TeleopTurtle::keyLoop()
...
+        RCLCPP_DEBUG(this->get_logger(), "DOWN");
         linear_ = -1.0;
         dirty = true;
         break;
+      case KEYCODE_S:
+        RCLCPP_DEBUG(this->get_logger(), "STOP");
+        linear_ = 0.0;
+        angular_ = 0.0;
+        dirty = true;
+        break;
```
Comment (dirk-thomas) [inline turtlesim/tutorials/teleop_turtle_key.cpp]: Please do not introduce new features in the same PR which performs the porting. Please contribute that separate to the default branch.

File: turtlesim/CMakeLists.txt [hunk 14]
```diff
@@ -1,55 +1,110 @@
...
+
+if(BUILD_TESTING)
+  find_package(ament_lint_auto REQUIRED)
+  ament_lint_auto_find_test_dependencies()
 
-install(TARGETS turtlesim_node turtle_teleop_key draw_square mimic
-  RUNTIME DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION})
+  find_package(ament_cmake_pytest REQUIRED)
+  find_package(launch_testing_ament_cmake REQUIRED)
+  find_package(rmw_implementation_cmake REQUIRED)
```
Comment (dirk-thomas) [inline turtlesim/CMakeLists.txt]: These three lines seem to do nothing?

File: turtlesim/CMakeLists.txt [hunk 15]
```diff
@@ -1,55 +1,110 @@
...
+install(DIRECTORY 
+  images
+  launch
+  DESTINATION share/${PROJECT_NAME}
+  # FILES_MATCHING PATTERN "*.png" PATTERN "*.svg"
+)
+
+if(BUILD_TESTING)
+  find_package(ament_lint_auto REQUIRED)
+  ament_lint_auto_find_test_dependencies()
```
Comment (dirk-thomas) [inline turtlesim/CMakeLists.txt]: Using the ROS 2 linters might no be feasible if we want the code to stay close to how it is formatted on the ROS 1 branch.

**Review by dirk-thomas — COMMENTED** [review]
I added some comments inline but didn't go through the whole patch for now. The general feedback:

* Please avoid any unnecessary changes in this patch (including style changes, refactorings, etc.). Only the bare minimum to convert from ROS 1 to ROS 2. The reason is to keep the `dashing-devel` branch as close to the `melodic-devel` branch as possible. Otherwise maintaining both branches moving forward will be much more effort.
* Any changes unrelated to the porting should be done in separate PRs. Preferably against the default branch from where they can be cherry-picked into the ROS 2 branch. Maybe consider getting them reviewed and merged first.

File: turtlesim/CMakeLists.txt (hunk 3 above)
Comment (routiful) [reply turtlesim/CMakeLists.txt]: I deleted white space

File: turtlesim/CMakeLists.txt (hunk 4 above)
Comment (routiful) [reply turtlesim/CMakeLists.txt]: Accepted

File: turtlesim/CMakeLists.txt (hunk 5 above)
Comment (routiful) [reply turtlesim/CMakeLists.txt]: I have deleted `launch`

File: turtlesim/include/turtlesim/turtle.h (hunk 6 above)
Comment (routiful) [reply turtlesim/include/turtlesim/turtle.h]: Accepted

File: turtlesim/CMakeLists.txt (hunk 5 above)
Comment (routiful) [reply turtlesim/CMakeLists.txt]: If it not installed, However, I can't launch `multisim.launch.py`.

File: turtlesim/include/turtlesim/turtle.h (hunk 7 above)
Comment (routiful) [reply turtlesim/include/turtlesim/turtle.h]: It makes more intuitive than float. But to keep original code, I have replaced it to qreal.

File: turtlesim/include/turtlesim/turtle.h (hunk 8 above)
Comment (routiful) [reply turtlesim/include/turtlesim/turtle.h]: Accepted

File: turtlesim/include/turtlesim/turtle.h (hunk 9 above)
Comment (routiful) [reply turtlesim/include/turtlesim/turtle.h]: Accepted

File: turtlesim/include/turtlesim/turtle_frame.h (hunk 10 above)
Comment (routiful) [reply turtlesim/include/turtlesim/turtle_frame.h:76]: Because lambda is recommended style in ROS2

File: turtlesim/src/turtle_frame.cpp (hunk 11 above)
Comment (routiful) [reply turtlesim/src/turtle_frame.cpp]: Accepted

File: turtlesim/tutorials/teleop_turtle_key.cpp (hunk 13 above)
Comment (routiful) [reply turtlesim/tutorials/teleop_turtle_key.cpp]: Accepted

File: turtlesim/CMakeLists.txt (hunk 14 above)
Comment (routiful) [reply turtlesim/CMakeLists.txt]: Yes, I have deleted it.

**Comment by routiful** [comment]
@dirk-thomas 
Thank you for your review. I almost have accepted your required.
I have tested this in dashing-release.

File: turtlesim/include/turtlesim/turtle.h [hunk 16]
```diff
@@ -56,60 +58,58 @@ namespace turtlesim
...
 
-  ros::NodeHandle nh_;
+  rclcpp::Node::SharedPtr nh_;
 
   QImage turtle_image_;
   QImage turtle_rotated_image_;
 
   QPointF pos_;
-  qreal orient_;
+  float orient_;
```
Comment (dirk-thomas) [inline turtlesim/include/turtlesim/turtle.h]: These types don't have to change either. Same as https://github.com/ros/ros_tutorials/pull/53#discussion_r292097587

**Review by dirk-thomas — COMMENTED** [review]
I added another comment and some previous comments are still pending.

The main goal is to keep the diff as minimal as possible. So if a change isn't required to make it work with ROS 2 it shouldn't be part of this PR.

File: turtlesim/include/turtlesim/turtle_frame.h (hunk 10 above)
Comment (dirk-thomas) [reply turtlesim/include/turtlesim/turtle_frame.h:76]: > Because lambda is recommended style in ROS2

Can you please post a link where you have read that.

Anyway in this case keeping the diff minimal is more important. So please keep the callback methods as they were before.

File: turtlesim/CMakeLists.txt (hunk 2 above)
Comment (routiful) [reply turtlesim/CMakeLists.txt]: I've removed trailing spaces as possible to compare original code

File: turtlesim/src/turtle_frame.cpp (hunk 12 above)
Comment (routiful) [reply turtlesim/src/turtle_frame.cpp]: I've deleted `// msec`

File: turtlesim/include/turtlesim/turtle.h (hunk 16 above)
Comment (routiful) [reply turtlesim/include/turtlesim/turtle.h]: I've replaced float to qreal in turtle.h and turtle.cpp

File: turtlesim/CMakeLists.txt (hunk 2 above)
Comment (dirk-thomas) [reply turtlesim/CMakeLists.txt]: Please recheck your changes. There are still several line with trailing spaces.

File: turtlesim/include/turtlesim/turtle_frame.h (hunk 10 above)
Comment (routiful) [reply turtlesim/include/turtlesim/turtle_frame.h:76]: > Can you please post a link where you have read that.

I didn't read a post. I thought demos repo in ros2 is developed by recommended style. 
Because server in demos represents lambdas (https://github.com/ros2/demos/blob/df5ec3a477c683bbaf9b3a2ec064680a0d36829d/demo_nodes_cpp/src/services/add_two_ints_server.cpp#L42)

But to keep the original code I've modified the code.

File: turtlesim/CMakeLists.txt (hunk 2 above)
Comment (routiful) [reply turtlesim/CMakeLists.txt]: Sorry, I've rechecked

Comment (dirk-thomas) [reply turtlesim/CMakeLists.txt]: There are still trailing spaces remaining. One on the specific line this comment is on. Anyway I went ahead and removed them myself: 335cbe6d8d71addc51de60a4766b8e622b1e01f2

I can only recommend to use an editor which is capable of highlighting such trailing space (beside other linters) and/or use a tool which is able to flag / remove them.

File: turtlesim/src/turtle_frame.cpp [hunk 17]
```diff
@@ -246,26 +248,26 @@ void TurtleFrame::updateTurtles()
   if (modified)
   {
     update();
-  }
-
+  }
```
Comment (dirk-thomas) [inline turtlesim/src/turtle_frame.cpp]: Unnecessary diff.

File: turtlesim/src/turtle_frame.cpp [hunk 18]
```diff
@@ -246,26 +248,26 @@ void TurtleFrame::updateTurtles()
...
 }
 
-
-bool TurtleFrame::clearCallback(std_srvs::Empty::Request&, std_srvs::Empty::Response&)
+bool TurtleFrame::clearCallback(const std::shared_ptr<std_srvs::srv::Empty::Request>, std::shared_ptr<std_srvs::srv::Empty::Response>)
 {
-  ROS_INFO("Clearing turtlesim.");
+  RCLCPP_INFO(nh_->get_logger(), "Clearing turtlesim.");
   clear();
+
```
Comment (dirk-thomas) [inline turtlesim/src/turtle_frame.cpp]: Unnecessary diff.

File: turtlesim/src/turtle_frame.cpp [hunk 19]
```diff
@@ -246,26 +248,26 @@ void TurtleFrame::updateTurtles()
...
-bool TurtleFrame::resetCallback(std_srvs::Empty::Request&, std_srvs::Empty::Response&)
+bool TurtleFrame::resetCallback(const std::shared_ptr<std_srvs::srv::Empty::Request>, std::shared_ptr<std_srvs::srv::Empty::Response>)
 {
-  ROS_INFO("Resetting turtlesim.");
+  RCLCPP_INFO(nh_->get_logger(), "Resetting turtlesim.");
   turtles_.clear();
   id_counter_ = 0;
   spawnTurtle("", width_in_meters_ / 2.0, height_in_meters_ / 2.0, 0);
   clear();
+
```
Comment (dirk-thomas) [inline turtlesim/src/turtle_frame.cpp]: Unnecessary diff.

File: turtlesim/src/turtlesim.cpp [hunk 20]
```diff
@@ -29,34 +29,37 @@
...
   int exec()
   {
-    turtlesim::TurtleFrame frame;
-    frame.show();
-
     return QApplication::exec();
   }
+
+private:
+	std::shared_ptr<turtlesim::TurtleFrame> frame_;
```
Comment (dirk-thomas) [inline turtlesim/src/turtlesim.cpp]: Why was the frame changed from a local variable in `exec` to a member variable? If there is no reason please revert.

File: turtlesim/tutorials/draw_square.cpp [hunk 21]
```diff
@@ -1,163 +1,230 @@
...
 };
 
-State g_state = FORWARD;
-State g_last_state = FORWARD;
-bool g_first_goal_set = false;
-
-#define PI 3.141592
-
-void poseCallback(const turtlesim::PoseConstPtr& pose)
+class DrawSquare : public rclcpp::Node
```
Comment (dirk-thomas) [inline turtlesim/tutorials/draw_square.cpp]: As of here the patch contains still largely unnecessary changes. Please keep the patch as close as possible to the upstream branch.

This applies to pretty much everything below.

**Comment by dirk-thomas** [comment]
(I will be out of the office for the next month. So it is unlikely that you will get further feedback until I am back mid July.)

File: turtlesim/src/turtle_frame.cpp (hunk 17 above)
Comment (routiful) [reply turtlesim/src/turtle_frame.cpp]: Accepted

File: turtlesim/src/turtle_frame.cpp (hunk 18 above)
Comment (routiful) [reply turtlesim/src/turtle_frame.cpp]: Accepted

File: turtlesim/src/turtle_frame.cpp (hunk 19 above)
Comment (routiful) [reply turtlesim/src/turtle_frame.cpp]: Accepted

File: turtlesim/src/turtlesim.cpp (hunk 20 above)
Comment (routiful) [reply turtlesim/src/turtlesim.cpp]: I've reverted

File: turtlesim/tutorials/draw_square.cpp (hunk 21 above)
Comment (routiful) [reply turtlesim/tutorials/draw_square.cpp]: Accepted

**Comment by routiful** [comment]
@dirk-thomas I've done to revert original code and delete some unnecessary diff.

**Comment by routiful** [comment]
@dirk-thomas Carefully Ping. Any comments?

File: turtlesim/src/turtle.cpp [hunk 22]
```diff
@@ -51,52 +50,55 @@ Turtle::Turtle(const ros::NodeHandle& nh, const QImage& turtle_image, const QPoi
...
   pen_.setWidth(3);
 
-  velocity_sub_ = nh_.subscribe("cmd_vel", 1, &Turtle::velocityCallback, this);
-  pose_pub_ = nh_.advertise<Pose>("pose", 1);
-  color_pub_ = nh_.advertise<Color>("color_sensor", 1);
-  set_pen_srv_ = nh_.advertiseService("set_pen", &Turtle::setPenCallback, this);
-  teleport_relative_srv_ = nh_.advertiseService("teleport_relative", &Turtle::teleportRelativeCallback, this);
-  teleport_absolute_srv_ = nh_.advertiseService("teleport_absolute", &Turtle::teleportAbsoluteCallback, this);
+  nh_ = node_handle;
+  last_command_time_ = nh_->now();
```
Comment (dirk-thomas) [inline turtlesim/src/turtle.cpp]: Can you describe why this line was added? It doesn't seem to exist before this patch.

**Review by dirk-thomas — COMMENTED** [review]
Can you please enable the flag for allowing me to push changes to your branch. Then we can keep iterating on this PR - otherwise I have to create a new PR with my changes to this branch. Thanks.

File: turtlesim/src/turtle.cpp (hunk 22 above)
Comment (routiful) [reply turtlesim/src/turtle.cpp]: This line is unneccesary. I've deleted it.

**Comment by routiful** [comment]
@dirk-thomas 
> Can you please enable the flag for allowing me to push changes to your branch. 

I don't understand which means of `flag`. Could you describe to me more specific??

**Comment by dirk-thomas** [comment]
> I don't understand which means of `flag`. Could you describe to me more specific??

https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by routiful** [comment]
@dirk-thomas 
Thanks. I'v enabled that check box.

**Comment by dirk-thomas** [comment]
CI confirms that the package build successfully on Linux / macOS (I doesn't have any tests):

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8109/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4106/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6607/)

On Windows it fails to build atm but that can be addressed in a separate PR:

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7989/)

Thank you for iterating on this even though it took me a while to get back at it.
